Add 3D convolution AI heuristics implementation

JH-Leon-KIM-AMD opened the pull request:
@amd-bartgips , @vpietila-amd , @amd-ahyttine 
3D tunanet code PR

- Add Gfx942Model_3D class for 3D convolution predictions
- Implement metadata_3d.cpp for 3D feature extraction  
- Move common AI heuristics code to ai_heuristics.hpp for better organization
- Functionalization for reusing and improving inefficient memory cache logic
- Add trained 3D TunaNet models for gfx942
- Update CMakeLists.txt for 3D components
- conflict test done with Bart latest commit

Test script

```
export MIOPEN_LOG_LEVEL=4
export MIOPEN_DEBUG_ENABLE_AI_IMMED_MODE_FALLBACK=1
export MIOPEN_ENABLE_AI_IMMED_MODE_FALLBACK=1
export MIOPEN_DISABLE_PERFDB=1
export MIOPEN_DISABLE_USERDB=1
export MIOPEN_DISABLE_FIND_DB=1
export MIOPEN_DISABLE_CACHE=1
export MIOPEN_DEBUG_CONV_IMMED_FALLBACK=1
export MIOPEN_DEBUG_FORCE_IMMED_MODE_FALLBACK=1
#3D convolution
./bin/MIOpenDriver conv -F 1 -n 1 -c 4 -k 8 -H 8 -W 8 -! 8 -y 3 -x 3 -@ 3 -_ 3 -t fp32 -V 0
```

JH-Leon-KIM-AMD committed:
Add 3D convolution AI heuristics implementation

  - Add Gfx942Model_3D class for 3D convolution predictions
  - Implement metadata_3d.cpp for 3D feature extraction
  - Add trained 3D TunaNet models for gfx942
  - Update CMakeLists.txt for 3D components

amd-bartgips commented on src/include/miopen/conv/heuristics/ai_heuristics.hpp:
```diff
@@ -42,9 +42,59 @@
 #include <miopen/any_solver.hpp>
 #include <miopen/filesystem.hpp>
 #include <miopen/anyramdb.hpp>
+#include <fstream>
```
I get a warning here when building:
```
/MIOpen_dev/src/include/miopen/conv/heuristics/ai_heuristics.hpp:45:1: warning: duplicate include [readability-duplicate-include]
   44 | #include <miopen/anyramdb.hpp>
      |                               
   45 | #include <fstream>
      | ^~~~~~~~~~~~~~~~~~
```
Is this already included somewhere else?

vpietila-amd replied:
You could try placing `#pragma once ` as the first line of the header file. It should fix the problem as it ensures that header is included only once. Since you have the .cpp file that contains the implementation, it is anyway a good practice.

JH-Leon-KIM-AMD replied:
Yes it's redundant include.
Thank you for the tip (#pragma once), I will keep in mind.

amd-bartgips commented on src/include/miopen/conv/heuristics/ai_heuristics.hpp:
```diff
@@ -42,9 +42,59 @@
 
 namespace miopen {
 namespace ai {
+
```
I like this refactor to miopen::ai::common, makes sense to me.

amd-bartgips replied:
A question from a C++ beginner: Why would we place these function definitions in the .hpp file instead of the .cpp file?

vpietila-amd replied:
It seems that MIOpen wants have the option to distribute the library as a header-only library, where the user can compile the library with their own code. When the implementation is in header files, this is possible. If you would place the implementation to .cpp files, you would need to distribute compiled libraries. This create problems with compiler/runtime versions since users of MIOpen would need to have versions that are compatible with the compiled binaries. Header-only library is free of such dependencies.

vpietila-amd replied:
Actually, since you already have the .cpp file, I would move all implementations to the .cpp file. The hybrid is just confusing.

vpietila-amd replied:
So only declarations (signature) in the header file and the implementation in the .cpp file.

amd-bartgips replied:
Related to this: I seem to recall John Shumway saying something about being header-only in our meeting yesterday... or was that only about CK?

JH-Leon-KIM-AMD replied:
I created metadata_3d.cpp, which needed to use ReverseMap and LoadJSON functions that were originally in hpp file. To make these functions accessible to both ai_heuristics.cpp and metadata_3d.cpp, I moved them to a header file (.hpp).

amd-bartgips commented on src/include/miopen/conv/heuristics/ai_heuristics.hpp:
```diff
@@ -76,6 +126,215 @@ MIOPEN_INTERNALS_EXPORT std::vector<uint64_t> PredictSolver(const conv::ProblemD
+/**
+ * @brief 3D-specific metadata handler for TunaNet3D models
+ * 
+ * This class provides a simple interface for accessing 3D convolution metadata.
```
did you need to alter the metadata json structure at all? Or did you simply use the current file format that our python code spits out?

JH-Leon-KIM-AMD replied:
I used the JSON structure generated from the Python code as is.

amd-bartgips commented on src/conv/heuristics/ai_heuristics.cpp:
```diff
@@ -500,87 +475,346 @@ class Gfx942Model final : public Model
+    }
+    
+    bool IsProblemSupported(const conv::ProblemDescription& problem,
+                            const ExecutionContext& ctx) const override
```
I receive a warning when building, since ctx is not used, perhaps do:

```suggestion
                            [[maybe_unused]] const ExecutionContext& ctx) const override
```

vpietila-amd replied:
It is easier to simply write 
```
bool IsProblemSupported(const conv::ProblemDescription& problem,
                            const ExecutionContext&) const override
```

If you are not using the `ctx`. This should remove the compiler warning.

JH-Leon-KIM-AMD replied:
Thank you Bart and Ville, I corrected.

amd-bartgips commented on src/kernels/gfx942_3d.tn.model:
As Ville rightfully pointed out. Perhaps we should make these git-lfs files. Can you perhaps check if this is the case with the other tunanet files?

JH-Leon-KIM-AMD replied:
**Currently in Git LFS:**
  - Only *.kdb.bz2 files

  **NOT in Git LFS**
  - All *.tn.model files (TunaNet models) - ranging from 2KB to 1.1MB
  - All *.ktn.model files (Kernel Tuning models) - ranging from 827 bytes to
  1.1MB
  
  **If we need to track git-lfs for model files then**
  
```
  git lfs track "*.tn.model"
  git lfs track "*.ktn.model"
  git add .gitattributes
  git add src/kernels/*.tn.model src/kernels/*.ktn.model
  git commit -m "Move TunaNet models to Git LFS"
```

amd-bartgips commented:
When I run your commands suggested above it is better to set the log level higher:
```
export MIOPEN_LOG_LEVEL=6
export MIOPEN_DEBUG_ENABLE_AI_IMMED_MODE_FALLBACK=1
export MIOPEN_ENABLE_AI_IMMED_MODE_FALLBACK=1
export MIOPEN_DISABLE_PERFDB=1
export MIOPEN_DISABLE_USERDB=1
export MIOPEN_DISABLE_FIND_DB=1
export MIOPEN_DISABLE_CACHE=1
export MIOPEN_DEBUG_CONV_IMMED_FALLBACK=1
export MIOPEN_DEBUG_FORCE_IMMED_MODE_FALLBACK=1
#3D convolution
./bin/MIOpenDriver conv -F 1 -n 1 -c 4 -k 8 -H 8 -W 8 -! 8 -y 3 -x 3 -@ 3 -_ 3 -t fp32 -V 0
```

So now it's easier to see that your new stuff is properly being called :)

```
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 4-8-8-8-3x3x3-8-6-6-6-1-0x0x0-1x1x1-1x1x1-0-NCDHW-FP32-F in cache for file "/root/.config/miopen/gfx942130.HIP.3_5_0_b09b18fb8-dirty.ufdb.txt"
MIOpen(HIP): Info2 [FindRecord] Looking for key 4-8-8-8-3x3x3-8-6-6-6-1-0x0x0-1x1x1-1x1x1-0-NCDHW-FP32-F in file "/home/bartgips/code/MIOpen_dev/build/share/miopen/db/gfx942130.HIP.fdb.txt"
MIOpen(HIP): Info2 [Measure] Db::FindRecord time: 0.105818 ms
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 4-8-8-8-3x3x3-8-6-6-6-1-0x0x0-1x1x1-1x1x1-0-NCDHW-FP32-F in cache for file ":memory:3d_gfx942"
MIOpen(HIP): Info2 [Get3DModel] Get3DModel called for device: gfx942
MIOpen(HIP): Info2 [Metadata3D] Metadata3D loaded for arch: gfx942_3d, num_inputs=23, num_solvers=9
MIOpen(HIP): Info2 [Gfx942Model_3D] Gfx942Model_3D initialized
MIOpen(HIP): Info2 [Get3DModel] Successfully created 3D model for device: gfx942
MIOpen(HIP): Info2 [IsProblemSupported] 3D problem supported by Gfx942Model_3D
MIOpen(HIP): Info2 [PredictSolver] Evaluating 3D TunaNet
MIOpen(HIP): Info2 [ToFeatures] Gfx942Model_3D: Extracted 23 features
MIOpen(HIP): Info2 [Forward] Gfx942Model_3D: Extracted 23 features
...
MIOpen(HIP): Info2 [Forward] Gfx942Model_3D: TunaNet3D returned 9 predictions
MIOpen(HIP): Info2 [PredictSolver] 3D TunaNet Result: ConvHipImplicitGemm3DGroupFwdXdlops ID:138, GemmFwdRest ID:91, ConvHipImplicitGemm3DGroupBwdXdlops ID:141, ConvHipImplicitGemm3DGroupWrwXdlops ID:140, GemmFwd1x1_0_1 ID:88, GemmWrwUniversal ID:102, GemmBwdRest ID:97, GemmBwd1x1_stride1 ID:96, GemmWrw1x1_stride1 ID:101, 
MIOpen(HIP): Info2 [GetSolutionsFallback] Using TunaNet Fallback
MIOpen(HIP): Info2 [GetSolutionsFallback] maxSolutionCount = 1, available = 2
MIOpen(HIP): Info2 [GetSolutionsFallback] id: 138, algo: 5, time: 10, ws: 18688, name: ConvHipImplicitGemm3DGroupFwdXdlops
MIOpen(HIP): Info2 [GetSolutionsFallback] id: 91, algo: 0, time: 20, ws: 93312, name: GemmFwdRest
```

Any idea why it gives us two solutions here at the end? (ConvHipImplicitGemm3DGroupFwdXdlops, GemmFwdRest). Shouldn't the tunanet only give us one of them? Or is one of the two a hard-coded fallback?

Besides this, as I said yesterday: it would be nice to have explicit test functions for you specific functions :)

vpietila-amd commented on src/conv/heuristics/metadata_3d.cpp:
```diff
@@ -0,0 +1,316 @@
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
```
Add the correct year.

vpietila-amd commented on src/conv/heuristics/metadata_3d.cpp:
```diff
@@ -0,0 +1,316 @@
+    } catch (...) {
+        success = false;
+        return {};
+    }
```
Perhaps a better pattern is to return `std::optional<std::vector<std::string>>` where `std::nullopt`  indicates failure. This would apply to all `LoadXXX` methods. Then you don't need additional `success` variable. Generally, the functional style of programming where functions do not have side-effects (such as modifying value of an input variable) is safer.

JH-Leon-KIM-AMD replied:
I didn't know the std::optional. Thank you. I've fixed the code.

JH-Leon-KIM-AMD replied:
I will keep in mind "So only declarations (signature) in the header file and the implementation in the .cpp file."
Since this was existing code, I'll probably need to refactor it after merging it with Bart. If I modify the existing code, Bart's code will likely stop working.
Thank you!!

vpietila-amd reviewed (approved):
Looks good to me.

JH-Leon-KIM-AMD committed:
remove redundant hpp include on fstream"

JH-Leon-KIM-AMD committed:
updated unused ExecutionContext parameter.

JH-Leon-KIM-AMD committed:
refactor: update metadata_3d exception pattern to use std::optional

JH-Leon-KIM-AMD committed:
Add comprehensive gtest suite for 3D convolution AI heuristics

  - Add 13 unit tests covering 3D AI heuristics functionality
  - Test metadata loading, model creation, and solver predictions
  - Add MIOpenDriver equivalent test (BartTest) for real-world validation
  - Enhance logging to show solver predictions with scores
  - Verify correct solver selection for 3D convolution problems

amd-bartgips commented on src/conv/heuristics/metadata_3d.cpp:
```diff
@@ -0,0 +1,348 @@
+
+} // namespace conv3d
+} // namespace ai
+} // namespace miopen
```
```suggestion
} // namespace miopen

```
the formatter prefers a newline at the end of the file

amd-bartgips commented on src/conv/heuristics/metadata_3d.cpp:
```diff
@@ -0,0 +1,348 @@
+    // For now, return default values (could be enhanced to load from JSON stats)
+    // This is a simplified version that returns ones for std
+    return std::vector<float>(num_inputs, 1.0f);
+}
```
The compiler again complains that "arch" is not used here. What was the trick again to make this optional?

amd-bartgips commented on src/conv/heuristics/metadata_3d.cpp:
```diff
@@ -0,0 +1,348 @@
+    
+    // All components loaded successfully, now we can safely move the data
+    // We need to const_cast because the members are const
+    const_cast<std::vector<std::string>&>(features) = std::move(*features_opt);
+    const_cast<size_t&>(num_inputs) = *num_inputs_opt;
+    const_cast<size_t&>(num_outputs) = *num_outputs_opt;
+    const_cast<size_t&>(num_solvers) = *num_solvers_opt;
+    const_cast<std::unordered_map<size_t, std::string>&>(solver_map) = std::move(*solver_map_opt);
+    const_cast<std::vector<float>&>(features_mean) = std::move(*features_mean_opt);
+    const_cast<std::vector<float>&>(features_std) = std::move(*features_std_opt);
+    const_cast<std::unordered_map<std::string, int>&>(direction_encodings_3d) = std::move(*direction_encodings_opt);
+    const_cast<std::unordered_map<std::string, int>&>(precision_encodings_3d) = std::move(*precision_encodings_opt);
+    const_cast<std::unordered_map<std::string, int>&>(in_layout_encodings) = std::move(*in_layout_encodings_opt);
+    const_cast<std::unordered_map<std::string, int>&>(fil_layout_encodings) = std::move(*fil_layout_encodings_opt);
+    const_cast<std::unordered_map<std::string, int>&>(out_layout_encodings) = std::move(*out_layout_encodings_opt);
```
The compiler does not like these `const_cast` calls:
```
/home/bartgips/code/MIOpen_dev/src/conv/heuristics/metadata_3d.cpp:312:5: warning: do not use const_cast to remove const qualifier [cppcoreguidelines-pro-type-const-cast]
  312 |     const_cast<std::vector<std::string>&>(features)                  = std::move(*features_opt);
      |     ^
```
Copilot suggests:

> Remove the const qualifier from the member variables of your Metadata3D class.
> - Remove const from member variables that are assigned in the constructor.
> - Make them private if you want to restrict modification.
> - Provide const getter methods if needed.
> 

Which makes sense to me

amd-bartgips replied:
I've implemented co-pilot's suggestion

amd-bartgips commented on test/gtest/conv_ai_3d_heuristics.cpp:
I think perhaps this could be slightly renamed to be more accurate? I.e. this is only to test the "tunanet" part, right? So maybe name it something like that? Or something with "solver_selection". 
This is just to make sure that this does not test the heuristics related to kernel tuning/parameter selection.

amd-bartgips replied:
Otherwise, this all looks good! All tests pass on my fresh build 🥳

amd-bartgips committed:
Added newline to end of file + Clang formatting

amd-bartgips committed:
Removed const_cast calls for private member variables in the constructor of Metadata3D, since the compiler did not like this.

amd-bartgips committed:
Added [[maybe_unused]] to silence warnings of unused arch arguments

amd-bartgips committed:
automated formatting using Clang

amd-bartgips committed:
More Clang formatting

amd-bartgips commented:
Merging this, next step: #3923
